software/apt-lib.pl

```diff
@@ -471,6 +498,40 @@ sub list_package_repos
 			$repo->{'id'} = $repo->{'url'}.$repo->{'name'};
 			push(@rv, $repo);
 			}
+		# New Ubuntu-style repos (using .sources files)
+		elsif ($f =~ /\.sources$/) {
+			if ($l =~ /^([\w\-]+):\s*(.+)$/) {
+				my ($key, $value) = ($1, $2);
+				if ($key eq 'Types') {
+					@types = split(/\s+/, $value);
+					}
+				elsif ($key eq 'URIs') {
+					$repo{'url'} = $value;
+					}
+				elsif ($key eq 'Suites') {
+					$repo{'suites'} = [split(/\s+/, $value)];
+					}
+				elsif ($key eq 'Components') {
+					$repo{'comps'} = [split(/\s+/, $value)];
+					}
+				elsif ($key eq 'Signed-By') {
+					$repo{'signed_by'} = $value;
+					}
+				elsif ($key eq 'Enabled') {
+					$repo{'disabled'} =
+						(lc($value) eq 'no') ? 1 : 0;
+					}
+				}
+			elsif ($l =~ /^\s*$/ || $lnum == $#{$lref}) {
```

Comment:
I think it's better to add the repo to the list when the first line is seen, not when it's the last one or the end of the file.

Reply:
We consider it a separator because the repository consists of a block (stanzas) of interconnected values, each divided by a new line.

https://manpages.debian.org/bookworm/dpkg-dev/deb822.5.en.html

That's why I add the repository only after the entire stanza has been read.

Reply:
Is this still going to work if there's no blank line at the end of the file? Seems like this branch won't be taken if the statement `if ($l =~ /^([\w\-]+):\s*(.+)$/) {` above matches.

Reply:
Oh, excellent catch! Thanks, I’ve just fixed that in this patch https://github.com/webmin/webmin/pull/2284/commits/c199cffb3d763a6f041437f54b0fd8b45fb8d062.